frontend/utils/distribution/group-income-distribution.test.js

```diff
@@ -0,0 +1,1852 @@
+/* eslint-env mocha */
+
+// For rapid development, run these tests with:
+// ./node_modules/.bin/mocha -w -R min --require Gruntfile.js frontend/utils/distribution/group-income-distribution.test.js
+
+import should from 'should'
+import { groupIncomeDistributionLogic, groupIncomeDistributionAdjustFirstLogic, groupIncomeDistributionNewLogic, dataToEvents } from './group-income-distribution.js'
+
+function unfoldParameters (json) {
+  const allPayments = {}
+  const paymentsFrom = {}
+
+  let paymentCount = 0
+  for (const p in json.payments) {
+    const payment = json.payments[p]
+    paymentCount++
+    const hash = 'paymentHash' + paymentCount
+    allPayments[hash] = {
+      'data': {
+        amount: payment.amount
+      }
+    }
+    if (!paymentsFrom[payment.from]) paymentsFrom[payment.from] = {}
+    if (!paymentsFrom[payment.from][payment.to]) paymentsFrom[payment.from][payment.to] = []
+    paymentsFrom[payment.from][payment.to].push(hash)
+  }
+
+  const monthlyPayments = {}
+  monthlyPayments[json.monthstamp] = {}
+  monthlyPayments[json.monthstamp].paymentsFrom = paymentsFrom
+
+  return {
+    getters:
+    {
+      'groupProfiles': json.profiles,
+      'currentGroupState': {
+        'payments': allPayments
+      },
+      'monthlyPayments': monthlyPayments,
+      'groupSettings': {
+        'mincomeAmount': json.mincome
+      }
+    },
+    monthstamp: json.monthstamp,
+    adjusted: json.adjusted
+  }
+}
+
+describe('Chunk 0: Adjustment Tests',
+  function () {
+    it('SCENARIO 1: can distribute income evenly with two users',
+      // has no effect for adjustment when there are no payments
+      // mincome: 12
+      // u1: pledge 10$
+      // u2: Income 10$ (a.k.a needs $2)
+      // Expected: [{ amount: 2, from: 'u1', to: 'u2' }]
+      function () {
+        const dist = groupIncomeDistributionAdjustFirstLogic(unfoldParameters({
+          'profiles': {
+            'u1': {
+              'incomeDetailsType': 'pledgeAmount',
+              'pledgeAmount': 10
+            },
+            'u2': {
+              'incomeDetailsType': 'incomeAmount',
+              'incomeAmount': 10
+            }
+          },
+          'payments': [],
+          'mincome': 12,
+          'monthstamp': '2020-11',
+          'adjusted': true
+        }))
```

Comment:
I think this can be simplified to make it even more readable. Here are some suggestions:

- Use defaults so that you can leave out data. For example, if the `payments` is empty (as it is here), don't require it to be included. If the `monthstamp` is always the same for the purpose of the tests, fill it in with a default value so that it too doesn't need to be specified. If `adjusted: true` is the value for that most of the time, then have that be the default and only specify it when it changes. DRY!
- Anywhere you have things spread across multiple lines, where they could be specified on a single line within 70 columns width, make them a single line (e.g. `u2: { 'incomeDetailsType': 'incomeAmount', 'incomeAmount': 10 }`)
- Anywhere something can be rewritten in a simpler way, do it (e.g. `u2: { need: 2 }` instead of the previous version)

Then these tests will read like a children's book and will be both super easy to write and read and understand.

Reply:
Copy/pasting from our call:

Scenario 4 TODOs:

1. Split Scenario 4.0 into two scenarios:
  - New Scenario 4.0:
    ```js
    expect({
      mincome: $1000
      u1: have: $100
      u2: need: $50
      u3: need: $1000
      payments: [{ u1: { u3: 50 }}]
    }).toEqual({
      // TODO: fill this in with the TODOs that appear
      //       in the UI for all users. (e.g. who pays
      //       who how much.)
    })
    ```
  - New Scenario 4.1:
    ```js
    expect({
      mincome: $1000
      u1: have: $50
      u2: need: $50
      u3: need: $1000
      payments: [{ u1: { u3: 50 }}]
    }).toEqual({
      // TODO: fill this in with the TODOs that appear
      //       in the UI for all users. (e.g. who pays
      //       who how much.)
      //       
      //       Note, u1 is pleding $50, and there is an
      //       existing payment from u1 of $50, so the
      //       TODO for u1 shouldn't exist anymore.
    })
    ```
  - New Scenario 4.2:
  ```js
  expect({
    mincome: $1000
    u1: have: $50
    u2: need: $50
    u3: need: $1000
    u4: have: $150
    payments: [{ u1: { u3: 50 }}]
  }).toEqual({
    // NOTE: we are only showing what Sandrina wrote
    //       here as is expected for u4, and we are
    //       not including the values (that in the real
    //       tests should be here) for the rest of the users
    // ... [snip] ...
    u4: { u2: 50, u3: 50 } // NOTE: the output will be
                           //       look more complicated
                           //       i'm just simplifying
                           //       it's ok to include the
                           //       complicated expected
                           //       output, which will look
                           //       something like:
                           //       
                           //       { from: 'u4', to: 'u2',
                           //        amount: 50 }
    // ... [snip] ...
  })
  ```